Cache fix

serbanghita opened the pull request:
- Kept the Cache implementation to be only PSR-16 compatible.
- Removed PSR-6 Cache implementation attempt.

Potential breaking changes.
Fixing #984, #979, #978


Backstory:

When I implemented cache for MobileDetect class I looked over PSR-6 and PSR-16 and I have combined both implementations and that resulted in ambiguous return values for Cache get or interfaces ambiguity (see #982).
I now looked over current various caching libs implementations in Github and saw that PSR-16 aka `psr/simple-cache` is the way to go.

serbanghita committed:
Breaking changes.
Fixing #984, #979, #978
Kept the Cache implementation to be only PSR-16 compatible.

serbanghita commented:
@ArchBlood @stof fyi

ArchBlood commented:
Just my two cents;

### 1. **Key Validation**
**PSR-16 Requirement:**
> Keys must be strings of at most 64 characters, containing only A-Z, a-z, 0-9, _, and . (dot).

**Fix:** Update `checkKey()` method to follow the spec:
```php
if (!is_string($key) || $key === '' || !preg_match('/^[A-Za-z0-9_.]{1,64}$/', $key)) {
    throw new InvalidArgumentException("Invalid key: '$key'. Must match /^[A-Za-z0-9_.]{1,64}$/");
}
```

---

### 2. **`has()` method logic**
**Issue:** Using `get()` can return the fallback default (`null`), which could falsely indicate a cache miss.

**Fix:** Check for key existence and TTL independently:
```php
public function has(string $key): bool
{
    $key = $this->checkKey($key);

    return isset($this->cache[$key]) &&
        ($this->cache[$key]['ttl'] === null || $this->cache[$key]['ttl'] > time());
}
```

---

### 3. **`fromIterable()` Validation**
**PSR-16 Requires:** `getMultiple()`, `setMultiple()`, and `deleteMultiple()` MUST validate input as array or Traversable, and validate **each key**.

**Fix:** Call `checkKey()` inside `getMultiple()`, `setMultiple()`, and `deleteMultiple()` for each key:
```php
foreach ($this->fromIterable($keys) as $key) {
    $key = $this->checkKey($key);
    ...
}
```

---

### 4. **Avoid Unused `checkKeyArray()`**
If you're not using `checkKeyArray()`, it's safe to remove it.

---

### 5. **PSR-16 Exceptions**
You should throw `Psr\SimpleCache\InvalidArgumentException`, not PHP’s built-in one.

**Fix:** Create a custom exception that implements the correct interface:
```php
use Psr\SimpleCache\InvalidArgumentException as PsrInvalidArgumentException;

class CacheException extends \InvalidArgumentException implements PsrInvalidArgumentException {}
```
And use it in `checkKey()`:
```php
throw new CacheException("...");
```

---

## ✅ Optional: Improvements

### ➕ Add expiration check abstraction
Extract TTL check to its own method:
```php
protected function isExpired(?int $ttl): bool
{
    return $ttl !== null && $ttl <= time();
}
```

---

## ✅ Summary of Required Fixes

| Area                     | Fix Required? | Description |
|--------------------------|----------------|-------------|
| Key Validation           | ✅             | Enforce PSR-compliant key rules |
| TTL Expiry Handling      | ✅             | `has()` must not rely on `get()` |
| Iterable Key Checks      | ✅             | Validate all keys in `*Multiple()` methods |
| Custom PSR Exception     | ✅             | Throw correct PSR exception |
| Cleanup Unused Methods   | ⚠️             | Remove `checkKeyArray()` if unused |

> Note: Not sure on the use case for why changing methods from `public` to `protected` is needed here.

serbanghita commented:
> ### 1. **Key Validation**

:+1: agree

> 
> ### 2. **`has()` method logic**
> **Issue:** Using `get()` can return the fallback default (`null`), which could falsely indicate a cache miss.
> 
> **Fix:** Check for key existence and TTL independently:
> 
> ```
> public function has(string $key): bool
> {
>     $key = $this->checkKey($key);
> 
>     return isset($this->cache[$key]) &&
>         ($this->cache[$key]['ttl'] === null || $this->cache[$key]['ttl'] > time());
> }
> ```

:+1: 

The docs are kinda ambiguous: `Determines whether an item is present in the cache.`
and this leads me to the fact that item MIGHT be present in the cache but not necessarily valid. But since `has()` only returns `boolean` then I deduce that it HAS to check for `ttl` as well.

I think you are correct, because you code also satisfies: `@throws \Psr\SimpleCache\InvalidArgumentException
     *   MUST be thrown if the $key string is not a legal value.`

Also `It is recommended that has() is only to be used for cache warming type purposes`. So cache warming should always know if an item is expired or not.

> 
> ### 3. **`fromIterable()` Validation**
> **PSR-16 Requires:** `getMultiple()`, `setMultiple()`, and `deleteMultiple()` MUST validate input as array or Traversable, and validate **each key**.
> 
> **Fix:** Call `checkKey()` inside `getMultiple()`, `setMultiple()`, and `deleteMultiple()` for each key:
> 
> ```
> foreach ($this->fromIterable($keys) as $key) {
>     $key = $this->checkKey($key);
>     ...
> }
> ```

:+1: 

Looks like `getMultiple`, `setMultiple` and `deleteMultiple` have to throw `\Psr\SimpleCache\InvalidArgumentException` if key is invalid.


> ### 4. **Avoid Unused `checkKeyArray()`**
> If you're not using `checkKeyArray()`, it's safe to remove it.

:+1:


> ### 5. **PSR-16 Exceptions**
> You should throw `Psr\SimpleCache\InvalidArgumentException`, not PHP’s built-in one.

:+1: 

> **Fix:** Create a custom exception that implements the correct interface:
> 
> ```
> use Psr\SimpleCache\InvalidArgumentException as PsrInvalidArgumentException;
> 
> class CacheException extends \InvalidArgumentException implements PsrInvalidArgumentException {}
> ```
> 
> And use it in `checkKey()`:
> 
> ```
> throw new CacheException("...");
> ```

That would be clean, but I'm kinda afraid of not interfering with somebody implementing a PSR-16 compatible cache system. Most likely I'll use the alias in code so I can see clearly what exception I'm throwing.

> 
> ## ✅ Optional: Improvements
> ### ➕ Add expiration check abstraction
> Extract TTL check to its own method:
> 
> ```
> protected function isExpired(?int $ttl): bool
> {
>     return $ttl !== null && $ttl <= time();
> }
> ```

:+1: 

> 
> ## ✅ Summary of Required Fixes
> Area	Fix Required?	Description
> Key Validation	✅	Enforce PSR-compliant key rules
> TTL Expiry Handling	✅	`has()` must not rely on `get()`
> Iterable Key Checks	✅	Validate all keys in `*Multiple()` methods
> Custom PSR Exception	✅	Throw correct PSR exception
> Cleanup Unused Methods	⚠️	Remove `checkKeyArray()` if unused


> > Note: Not sure on the use case for why changing methods from `public` to `protected` is needed here.

I'm thinking that `Cache` class is basically a suggested in-memory solution, but typically you want to create your own based on Redis or other storage when you inject your own cache system. I don't think anybody cares to extend `Cache`.

ArchBlood commented:
> > > Note: Not sure on the use case for why changing methods from `public` to `protected` is needed here.
> 
> I'm thinking that `Cache` class is basically a suggested in-memory solution, but typically you want to create your own based on Redis or other storage when you inject your own cache system. I don't think anybody cares to extend `Cache`.

I was just afraid that changing these methods would affect the implementation of the interface itself. 🤔

serbanghita commented:
> I was just afraid that changing these methods would affect the implementation of the interface itself. 🤔

@ArchBlood I checked `Cache` so now every method from `Psr\SimpleCache\CacheInterface` is still `public`, but the helpers and the rest are `protected`. Not sure I understand this.

Btw great comments 🙇 , really made me throw a hard look over the entire cache related code. Will come back today with the code changes.

ArchBlood commented:
> > I was just afraid that changing these methods would affect the implementation of the interface itself. 🤔
> 
> @ArchBlood I checked `Cache` so now every method from `Psr\SimpleCache\CacheInterface` is still `public`, but the helpers and the rest are `protected`. Not sure I understand this.
> 
> Btw great comments 🙇 , really made me throw a hard look over the entire cache related code. Will come back today with the code changes.

I look forward to the changes, and I'm sure that everyone that uses this will appreciate all of the hard work as well!

stof commented:
> That would be clean, but I'm kinda afraid of not interfering with somebody implementing a PSR-16 compatible cache system. Most likely I'll use the alias in code so I can see clearly what exception I'm throwing.

your own exception class would be in your own namespace, so there is no interference.
throwing an exception that does not implement the PSR-16 interface would be what could interfere with error handling of code expecting PSR-16.

stof commented on src/Cache/Cache.php:
```diff
@@ -1,104 +1,195 @@
     }
 
-    public function deleteMultiple(iterable $keys): bool
+    /**  */
```
missing phpdoc:

```
@param string[] $keys
@return string[]
```

stof commented on src/Cache/Cache.php:
```diff
@@ -1,104 +1,195 @@
-        foreach ($values as $key => $value) {
-            $this->set($key, $value, $ttl);
+
+        if (empty($key)) {
```
`'0'` is a valid key in PSR-16, but this implementation rejects it (basically, don't use `empty()` on strings in PHP as it is weird)

ArchBlood replied:
> `'0'` is a valid key in PSR-16, but this implementation rejects it (basically, don't use `empty()` on strings in PHP as it is weird)

I was a little confused when I seen this as well, I think it was thought to be implied that `empty()` would return `0` but here returns `true` or `false` depending on the condition, although leaves room for improvements;

### Example
```php
/**
 * @throws \InvalidArgumentException
 */
protected function checkKey(string $key): string
{
    // Check if the key is empty or invalid
    if ($key === '' || $key === '0' || $key === null) {
        throw new InvalidArgumentException('Cache key is empty');
    }

    return $key;
}
```
But as I stated in https://github.com/serbanghita/Mobile-Detect/pull/986#issuecomment-2784437676 should really think about implementing `Psr\SimpleCache\InvalidArgumentException` or an extended exception for here instead of using PHP's built-in version. 🤔

stof replied:
The issue is that `empty('0')` returns `true` (and so your implementation rejects it, but `'0'` is not an empty string (and so is not invalid in PSR-16)

serbanghita replied:
@stof checking again and will write a test specifically for this 🙇

serbanghita replied:
@stof implementation of Cache has been changed. Test suite covers 100% of code. This case has been addressed

stof commented on src/Cache/Cache.php:
```diff
@@ -1,104 +1,195 @@
+            return iterator_to_array($data); // @codeCoverageIgnore
+        }
+
+        throw new InvalidArgumentException('invalid data');
```
This is dead code, as `iterable` is the union type `array|\Traversable`

stof commented on src/Cache/Cache.php:
```diff
@@ -1,104 +1,195 @@
+    /**
+     * Get all cache keys. Needed for testing purposes.
+     *
+     * @return array{string}
```
this should be `string[]` or `array<string>`. `array{string}` is an array shape with a single key 0 being a string, which is not what this method returns.

stof commented on src/Cache/Cache.php:
```diff
@@ -1,104 +1,195 @@
 
-    public function has(string $key): bool
+    /**
+     * Get all cache keys. Needed for testing purposes.
```
this should probably be tagged `@internal` if it is only for testing.

stof commented on src/Cache/Cache.php:
```diff
@@ -1,104 +1,195 @@
+    }
+
+    /**
+     * @param bool[]|int[] $booleans
```
why supporting an array of integers ? You always pass an array of booleans

stof commented on src/MobileDetect.php:
```diff
@@ -1433,6 +1428,8 @@ public function isMobile(): bool
                 return $result;
             }
+        } catch (CacheException $e) {
+            throw new MobileDetectException("Cache problem in isMobile(): {$e->getMessage()}");
```
this should probably attach `$e` as previous exception to make debugging easier.

stof commented on standalone/autoloader.php:
```diff
@@ -5,21 +5,13 @@
     $classMap = [
         // "mobiledetect/mobiledetectlib"
         "Detection\Cache\Cache" => $dir . "/../src/Cache/Cache.php",
-        "Detection\Cache\CacheException" => $dir . "/../src/Cache/CacheException.php",
```
shouldn't this be kept ?

stof commented on tests/CacheTest.php:
```diff
@@ -16,83 +14,84 @@ protected function setUp(): void
     public function testGetInvalidCacheKey()
     {
-        $this->expectException(CacheException::class);
+        $this->expectException(\InvalidArgumentException::class);
```
this should expect the `Psr\SimpleCache\InvalidArgumentException`

stof reviewed (commented):
You might want to add https://packagist.org/packages/cache/integration-tests in your dev dependencies to use this shared testsuite for PSR-16 (and PSR-6 but this is not relevant here) to ensure your implementation follows the PSR-16 rules properly.

Tynael commented on tests/CacheTest.php:
```diff
@@ -16,83 +14,84 @@ protected function setUp(): void
+    /**
+     * @throws InvalidArgumentException
+     */
+    public function testSetTtlAsInteger()
```
[Nice-to-have]
You use `TTL` in some methods while in some `Ttl`. It'd be nice to have some consistency.

Tynael commented on tests/CacheTest.php:
```diff
@@ -16,83 +14,84 @@ protected function setUp(): void
+     * @throws InvalidArgumentException
      */
-    public function testGetExpiresAfter()
+    public function testSetTtlAsNull()
```
Many of the methods are missing the return type hint. It'd be good to have some consistency.

serbanghita replied:
I don't use phpcs for tests because I need more liberty when writing them and less clutter (for now)

Tynael commented on composer.json:
```diff
@@ -15,15 +15,14 @@
         "friendsofphp/php-cs-fixer": "^v3.65.0",
         "phpunit/phpunit": "^9.6.18",
-        "squizlabs/php_codesniffer": "^3.11.1",
+        "squizlabs/php_codesniffer": "3.12.1",
```
I think it's ok to add `^` for dev dependencies. This way it accepts any new patch or minor versions, as long as it doesn't break backward compatibility.

```suggestion
        "squizlabs/php_codesniffer": "^3.12.1",
```

serbanghita replied:
done

Tynael commented on src/Cache/Cache.php:
```diff
@@ -1,104 +1,195 @@
+        throw new InvalidArgumentException('invalid data');
+    }
+
+    /**  */
```
Looks like it's missing phpdoc

serbanghita committed:
Cache is now PSR-16 compatible
Code review phase 1 addressed

serbanghita committed:
Lowered phpunit version to satisfy php 8.2

serbanghita committed:
Back to phpunit 9.x because of >=8.0 support

ArchBlood commented on src/Cache/Cache.php:
```diff
@@ -1,104 +1,189 @@
+    {
+        $key = $this->checkKey($key);
+
+        return isset($this->cache[$key]);
```
Your `has()` method only checks with `isset()` This will return `true` even if the item has expired.

✅ Fix: Update `has()` to check TTL:
```php
public function has(string $key): bool
{
    $this->checkKey($key);

    if (!isset($this->cache[$key])) {
        return false;
    }

    if ($this->cache[$key]['ttl'] === null || $this->cache[$key]['ttl'] > time()) {
        return true;
    }

    $this->deleteSingle($key);
    return false;
}
```

serbanghita replied:
@ArchBlood I thought about this but I checked the specs and it says nothing about cleaning up the cache record (like in `set` and `get`)

```
/**
     * Determines whether an item is present in the cache.
     *
     * NOTE: It is recommended that has() is only to be used for cache warming type purposes
     * and not to be used within your live applications operations for get/set, as this method
     * is subject to a race condition where your has() will return true and immediately after,
     * another script can remove it, making the state of your app out of date.
     *
     * @param string $key The cache item key.
     *
     * @return bool
     *
     * @throws \Psr\SimpleCache\InvalidArgumentException
     *   MUST be thrown if the $key string is not a legal value.
     */
```

serbanghita commented:
fyi: Still wip, even if all tests pass

> You might want to add https://packagist.org/packages/cache/integration-tests in your dev dependencies to use this shared testsuite for PSR-16 (and PSR-6 but this is not relevant here) to ensure your implementation follows the PSR-16 rules properly.

@stof  Thank you, checking this right now after I refactored `Cache` class, looks like it fails at all tests atm :dagger: 

![Screenshot from 2025-04-09 21-47-50](https://github.com/user-attachments/assets/ee214dd2-28bd-45b0-8bb2-49a1a284af53)

```
Failed asserting that exception of type "TypeError" matches expected exception "Psr\SimpleCache\InvalidArgumentException". Message was: "Detection\Cache\Cache::has(): Argument #1 ($key) must be of type string, array given, called in /home/sghita/work/Mobile-Detect/vendor/cache/integration-tests/src/SimpleCacheTest.php on line 541" at
```

I think these `TypeError`s are fine, because types were introduced in PHP 8.0 so the code doesn't expect to be passed other key values than `string` 

@Tynael thanks for the review, addressed all

serbanghita commented:
I've also found `testBasicUsageWithLongKey` from `vendor/cache/integration-tests/src/SimpleCacheTest.php` which sets a huge cache key but I think that's wrong because it's against the PSR-16 spec (`UTF-8 encoding and a length of up to 64 characters.`) and can cause [data too long issues](https://github.com/serbanghita/Mobile-Detect/issues/974#issuecomment-2531597903)

ArchBlood reviewed (commented):
### PSR-16 specifies:
- Methods like `delete()`, `clear()`, etc., MUST return `bool`. You already do this, but just ensure no exceptions are thrown unless it's about an invalid key.
- ✅ You're safe as long as `checkKey()` is the only method that throws.

### Minor Suggestions (Optional)
- Error messages from `checkKey()` can optionally match PSR’s less opinionated guidance. PSR only requires an exception for invalid keys; the format enforcement (regex) is yours — valid, but stricter than PSR.
- `checkReturn()` is fine but could be simplified with `in_array(false, $array, true) === false.`

serbanghita committed:
More code review comments addressed.

ArchBlood commented:
> fyi: Still wip, even if all tests pass
> 
> > You might want to add https://packagist.org/packages/cache/integration-tests in your dev dependencies to use this shared testsuite for PSR-16 (and PSR-6 but this is not relevant here) to ensure your implementation follows the PSR-16 rules properly.
> 
> @stof Thank you, checking this right now after I refactored `Cache` class, looks like it fails at all tests atm 🗡️
> 
> ![Screenshot from 2025-04-09 21-47-50](https://private-user-images.githubusercontent.com/1106849/431989255-ee214dd2-28bd-45b0-8bb2-49a1a284af53.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQyMjY2MzMsIm5iZiI6MTc0NDIyNjMzMywicGF0aCI6Ii8xMTA2ODQ5LzQzMTk4OTI1NS1lZTIxNGRkMi0yOGJkLTQ1YjAtOGJiMi00OWExYTI4NGFmNTMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MDlUMTkxODUzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YmE4MDZhNjQ2ZGEwNTY0ZDI5MzA3OTczODZkNmJkNzI2NTBmYWE4YjFiYWEwMWQ4YjVlYTBjM2EzYWE2YWRmZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.yvVoL6VPzhne6yijXPLM2NhZkR_CDdT5mgOhf0gvzHs)
> 
> ```
> Failed asserting that exception of type "TypeError" matches expected exception "Psr\SimpleCache\InvalidArgumentException". Message was: "Detection\Cache\Cache::has(): Argument #1 ($key) must be of type string, array given, called in /home/sghita/work/Mobile-Detect/vendor/cache/integration-tests/src/SimpleCacheTest.php on line 541" at
> ```
> 
> I think these `TypeError`s are fine, because types were introduced in PHP 8.0 so the code doesn't expect to be passed other key values than `string`
> 
> @Tynael thanks for the review, addressed all

You need to remove the scalar type hint `string` from the public methods, and instead validate the type manually in `checkKey()`. That way, you can throw the correct PSR-16 exception.

### Example `has()`:
```php
public function has($key): bool
{
    $this->checkKey($key);

    if (!isset($this->cache[$key])) {
        return false;
    }

    if ($this->cache[$key]['ttl'] === null || $this->cache[$key]['ttl'] > time()) {
        return true;
    }

    $this->deleteSingle($key);
    return false;
}
```
### Example `checkKey()`:
```php
protected function checkKey(mixed $key): string
{
    if (!is_string($key)) {
        throw new CacheInvalidArgumentException("Invalid key type: must be a string.");
    }

    if ($key === '' || !preg_match('/^[A-Za-z0-9_.]{1,64}$/', $key)) {
        throw new CacheInvalidArgumentException("Invalid key: '$key'. Must be alphanumeric, can contain _ and . and can be max 64 characters.");
    }

    return $key;
}
```

serbanghita commented:
@ArchBlood no way I'm removing scalar types, they are mentioned in the original interface

![Screenshot from 2025-04-09 22-26-25](https://github.com/user-attachments/assets/57e528aa-a55d-4c87-9224-03bf8a6a9916)

are you using AI-tools to check the changes?

serbanghita committed:
revert to phpbench 1.2 to support PHP >=8.0

ArchBlood commented:
> @ArchBlood no way I'm removing scalar types, they are mentioned in the original interface
> 
> ![Screenshot from 2025-04-09 22-26-25](https://private-user-images.githubusercontent.com/1106849/432003308-57e528aa-a55d-4c87-9224-03bf8a6a9916.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQyMjcxNzAsIm5iZiI6MTc0NDIyNjg3MCwicGF0aCI6Ii8xMTA2ODQ5LzQzMjAwMzMwOC01N2U1MjhhYS1hNTVkLTRjODctOTIyNC0wM2JmOGE2YTk5MTYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MDlUMTkyNzUwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YThiYjI3NmYzNjI2YTM3OTg4YmRiYTQ3YmM1ZmU3NDdjMmZmYTA0Nzc5MTBiZGE4NzJlZmY1ZTA1ODgxMDFiNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Z0IU45qUlcRBIJsxph2_SmUV6i5IR6PV19Vy8F3qrHc)
> 
> are you using AI-tools to check the changes?

I ran a check through GitHub on a private action in a HumHub-specific container to identify any issues. I'm in complete agreement if you'd prefer not removing scalar types and relying on using `checkKey()` to throw an exception for strings, should keep it in compliance either way tbh.

serbanghita committed:
More CR addressed
Added MobileDetectExceptionCode class (enums not supported by 8.0)

serbanghita committed:
Added test for custom class implementing PSR-16 interface

serbanghita committed:
Covering 100% all Cache class

serbanghita commented:
Addressed all your comments @stof @Tynael :bow:

serbanghita committed:
Updated website deploy code
